nx/lib/nx.ex

```diff
@@ -12096,27 +12157,72 @@ defmodule Nx do
           [4, 9, 10]
         ]
       >
+
+  ## Vectorized tensors
+
+  The both tensor to be sliced and the slices can be vectorized,
+  but indices must be non-vectorized.
+
+      iex> t = Nx.tensor([[1, 2, 3, 4], [5, 6, 7, 8]]) |> Nx.vectorize(:x)
+      iex> slice = Nx.tensor([[10, 20], [30, 40]]) |> Nx.vectorize(:y)
+      iex> Nx.put_slice(t, [2], slice)
+      #Nx.Tensor<
+        vectorized[x: 2][y: 2]
+        s64[4]
+        [
+          [
+            [1, 2, 10, 20],
+            [1, 2, 30, 40]
+          ],
+          [
+            [5, 6, 10, 20],
+            [5, 6, 30, 40]
+          ]
+        ]
+      >
+
   """
   @doc type: :indexed
   def put_slice(tensor, start_indices, slice) when is_list(start_indices) do
-    %T{shape: shape, names: names, type: type} = tensor = to_tensor(tensor)
-    %T{shape: slice_shape, names: slice_names, type: slice_type} = slice = to_tensor(slice)
+    [tensor, slice] = broadcast_vectors([tensor, slice])
 
-    Nx.Shared.raise_vectorized_not_implemented_yet(tensor, __ENV__.function)
-    Nx.Shared.raise_vectorized_not_implemented_yet(slice, __ENV__.function)
+    %T{vectorized_axes: vectorized_axes, shape: shape, names: names, type: type} = tensor
+    %T{shape: slice_shape, names: slice_names, type: slice_type} = slice
 
     output_type = binary_type(type, slice_type)
 
     start_indices = to_indices(start_indices)
 
-    {shape, names} = Nx.Shape.put_slice(shape, names, slice_shape, slice_names, start_indices)
+    {output_shape, output_names} =
+      Nx.Shape.put_slice(shape, names, slice_shape, slice_names, start_indices)
 
-    impl!(tensor).put_slice(
-      %{tensor | shape: shape, names: names, type: output_type},
-      tensor,
-      start_indices,
-      slice
-    )
+    offset = length(vectorized_axes)
+
+    start_indices = List.duplicate(0, offset) ++ start_indices
+
+    offset_shape = Keyword.values(vectorized_axes)
+
+    tensor = devectorize(tensor)
+    slice = devectorize(slice)
+
+    output_shape_devec =
+      if offset != 0 do
+        List.to_tuple(offset_shape ++ Tuple.to_list(output_shape))
+      else
+        output_shape
+      end
+
+    output_names = List.duplicate(nil, offset) ++ output_names
+
+    result =
+      impl!(tensor).put_slice(
+        %{tensor | shape: output_shape_devec, names: output_names, type: output_type},
+        tensor,
+        start_indices,
+        slice
+      )
```

Comment:
It feels we can share logic with slice but it is your call if it makes sense. :)

Reply:
I'll merge it as is. Even though there are common sections of the code, I think extracting the would make for less readable code because we'd have to intersperse the extracted call with the non-common sections